python/runtime/cudaq/distributed/mpi_comm_impl.cpp

```diff
@@ -0,0 +1,609 @@
+    auto mpiMod = py::module::import("mpi4py.MPI");
+    auto inplaceAddr = mpiMod.attr("IN_PLACE");
+    py::tuple recvBuf = py::make_tuple(packData(recvBuffer, count, dataType),
+                                       count, convertType(dataType));
+    auto pyComm = unpackMpiCommunicator(comm);
+
+    pyComm.attr("Allreduce")(inplaceAddr, recvBuf, convertType(opType));
+    return 0;
+  } catch (std::exception &e) {
+    std::cerr << "[mpi4py] Caught exception \"" << e.what() << "\"\n";
+    return 1;
+  }
+}
+
+/// @brief Wrapper of MPI_Allgather
+static int mpi_Allgather(const cudaqDistributedCommunicator_t *comm,
+                         const void *sendBuffer, void *recvBuffer,
+                         int32_t count, DataType dataType) {
+  try {
+    auto pyComm = unpackMpiCommunicator(comm);
+    const auto size = pyComm.attr("Get_size")().cast<int>();
+    pyComm.attr("Allgather")(packData(sendBuffer, count, dataType, true),
+                             packData(recvBuffer, size * count, dataType));
+    return 0;
+  } catch (std::exception &e) {
+    std::cerr << "[mpi4py] Caught exception \"" << e.what() << "\"\n";
+    return 1;
+  }
+}
+
+/// @brief Wrapper of MPI_Allgatherv
+static int mpi_AllgatherV(const cudaqDistributedCommunicator_t *comm,
+                          const void *sendBuf, int sendCount, void *recvBuf,
+                          const int *recvCounts, const int *displs,
+                          DataType dataType) {
+  try {
+    auto pyComm = unpackMpiCommunicator(comm);
+    const auto size = pyComm.attr("Get_size")().cast<int>();
+    py::array_t<int> pyRecvCounts(size);
+    py::array_t<int> pyDispls(size);
+    auto pyRecvCountsData = pyRecvCounts.mutable_unchecked<1>();
+    auto pyDisplsData = pyDispls.mutable_unchecked<1>();
+    for (int i = 0; i < size; ++i) {
+      pyRecvCountsData(i) = recvCounts[i];
+      pyDisplsData(i) = displs[i];
+    }
+    pyComm.attr("Allgatherv")(
+        packData(sendBuf, sendCount, dataType, true),
+        py::make_tuple(packData(recvBuf, size * sendCount, dataType),
+                       pyRecvCounts, pyDispls, convertType(dataType)));
+    return 0;
+  } catch (std::exception &e) {
+    std::cerr << "[mpi4py] Caught exception \"" << e.what() << "\"\n";
+    return 1;
+  }
+  return 0;
+}
+
+/// @brief Wrapper of MPI_Isend and track pending requests for synchronization
+static int mpi_SendAsync(const cudaqDistributedCommunicator_t *comm,
+                         const void *buf, int count, DataType dataType,
+                         int peer, int32_t tag) {
+  if (PendingRequest::g_requests[comm].nActiveRequests == 2)
+    return -1;
+  try {
+    auto pyComm = unpackMpiCommunicator(comm);
+    auto request =
+        pyComm.attr("Isend")(packData(buf, count, dataType, true), peer, tag);
+    PendingRequest::g_requests[comm]
+        .requests[PendingRequest::g_requests[comm].nActiveRequests] = request;
+    ++PendingRequest::g_requests[comm].nActiveRequests;
```

Comment:
Statements like this are not thread safe because multiple threads could be executing this statement simultaneously. If we don't intend for this API to be thread safe, then we need to make that clear in the docs (but I suspect it would be better to make them thread safe, at least for the C++ implementation). It might be worth getting others to weigh in.

Reply:
Agreed. I added a `scoped_lock` to prevent race conditions here. FWIW, in the use cases that we are looking to support, e.g., using this MPI interface as a proxy for [custatevec MPI Comm](https://github.com/NVIDIA/cuQuantum/blob/main/samples/custatevec/samples_mpi/mpicomm.c), multi-threading is currently not an issue and we don't directly expose them to end users as `cudaq::mpi` APIs.